examples/azure/client_factory.md

```diff
@@ -0,0 +1,133 @@
+
+This Environment structure has the following fields configured for each cloud environment:
+
+```golang
+ManagementPortalURL          
+PublishSettingsURL           
+ServiceManagementEndpoint    
+ResourceManagerEndpoint      
+ActiveDirectoryEndpoint      
+GalleryEndpoint              
+KeyVaultEndpoint             
+GraphEndpoint                
+ServiceBusEndpoint           
+BatchManagementEndpoint      
+StorageEndpointSuffix        
+SQLDatabaseDNSSuffix         
+TrafficManagerDNSSuffix      
+KeyVaultDNSSuffix            
+ServiceBusEndpointSuffix     
+ServiceManagementVMDNSSuffix 
+ResourceManagerVMDNSSuffix   
+ContainerRegistryDNSSuffix   
+CosmosDBDNSSuffix            
+TokenAudience                
+APIManagementHostNameSuffix  
+SynapseEndpointSuffix        
+ResourceIdentifiers          
+```
+
+Using these URI's, `client_factory` is able to instantiate and return properly configured SDK clients without module developers having to duplicate this code.
+
+## Configuration and Defaults
+
+To configure different cloud environments, we will use the same `AZURE_ENVIRONMENT` environment variable that the Go SDK uses. This can be set to one of the following values:
+
+|Value                      |Cloud Environment  |
+|---------------------------|-------------------|
+|"AzureChinaCloud"          |ChinaCloud         |
+|"AzureGermanCloud"         |GermanCloud        |
+|"AzurePublicCloud"         |PublicCloud        |
+|"AzureUSGovernmentCloud"   |USGovernmentCloud  |
+|"AzureStackCloud"          |Azure stack        |
+
+Note that when using the "AzureStackCloud" setting, you must also set the `AZURE_ENVIRONMENT_FILEPATH` variable to point to a JSON file containing your Azure Stack URI deatils.
+
+>NOTE: The default behavior of the `client_factory` is to use the AzurePublicCloud environment. This requires no work from the developer to configure, and ensures consistent behavior with the current SDK code.       
+
+## Usage Patterns
+
+Modules authors will interact with the `client_factory` through the `NewClientFactory` method on the `azure` package as shown below:
+
+```golang
+    // Get a new factory instance and get requested client
+	factory := NewClientFactory()
+	client, err := factory.GetClientE(VirtualMachinesClientType, subscriptionID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Type cast and verify
+	vmClient, ok := client.(compute.VirtualMachinesClient)
+	if !ok {
+		return nil, fmt.Errorf("Unable to convert client to type compute.VirtualMachinesClient")
+	}
+```
+
+The `ClientFactory` interface currently exposes the following methods - these will be built out more as more modules are developed and integrated:
+
+* `GetClientE(clientType ClientType, subscriptionID string)`: returns a client instance based on the ClientType passed, along with any error.
+* `GetClientBaseURIE(clientType ClientType, client interface{})`: returns the configured BaseURI for the given client, along with any error.
```

Comment:
Any reason not to expose the non-E versions of the methods (i.e., `GetClient` and `GetClientBaseURI`) that fail the test if there was any error?

Reply:
Removed these specific examples from the docs; however, the `client_factory` is not intended to be used by _test writers_; instead, it's in support of _module authors_ to correctly support multicloud environment details.